src/Tools/dotnet-monitor/CollectionRules/CollectionRuleService.cs

```diff
@@ -0,0 +1,178 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Monitoring.WebApi;
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Actions;
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Configuration;
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Options;
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Triggers;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules
+{
+    internal class CollectionRuleService : IAsyncDisposable
+    {
+        private readonly ActionListExecutor _actionListExecutor;
+        private readonly CancellationTokenSource _disposalTokenSource = new();
+        private readonly ILogger<CollectionRuleService> _logger;
+        private readonly IOptionsMonitor<CollectionRuleOptions> _optionsMonitor;
+        private readonly CollectionRulesConfigurationProvider _provider;
+        private readonly List<Task> _runTasks = new();
+        private readonly ICollectionRuleTriggerOperations _triggerOperations;
+
+        private bool _disposed;
+
+        public CollectionRuleService(
+            ILogger<CollectionRuleService> logger,
+            CollectionRulesConfigurationProvider provider,
+            ICollectionRuleTriggerOperations triggerOperations,
+            ActionListExecutor actionListExecutor,
+            IOptionsMonitor<CollectionRuleOptions> optionsMonitor
+            )
+        {
+            _actionListExecutor = actionListExecutor ?? throw new ArgumentNullException(nameof(actionListExecutor));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _provider = provider ?? throw new ArgumentNullException(nameof(provider));
+            _optionsMonitor = optionsMonitor ?? throw new ArgumentNullException(nameof(optionsMonitor));
+            _triggerOperations = triggerOperations ?? throw new ArgumentNullException(nameof(triggerOperations));
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            lock (_disposalTokenSource)
+            {
+                if (_disposed)
+                {
+                    return;
+                }
+                _disposed = true;
+            }
+
+            _disposalTokenSource.Cancel();
```

Comment:
nit: If you are cancelling in the dispose, you may need to guard Cancel

Reply:
We probably have a lot of places in our code that does this. I'll create an extension method to safely call Cancel e.g. `static void SafeCancel(this CancellationTokenSource source)`.